modules/dag-l0/src/main/scala/io/constellationnetwork/dag/l0/http/routes/DelegatedStakesRoutes.scala

```diff
@@ -140,6 +147,144 @@ final case class DelegatedStakesRoutes[F[_]: Async: Hasher](
+        .get(environment)
+        .pure[F]
+        .flatMap(Async[F].fromOption(_, new RuntimeException(s"Could not retrieve emission config for env: $environment")))
+
+      (_, _, totalDelegateStake, currentTotalSupply) <- processDelegations(lastSnapshotInfo)
+      latestDelegateRewardsNoCommission <- getLatestDelegateRewardTotal(lastSnapshot, lastSnapshotInfo)
+
+      currentPrice <- toAmount(getCurrentDagPrice(emConfig))
+      nextPrice <- getNextDagPrice(emConfig)
+
+      avgRewardAmount <- calculateAverageReward(latestDelegateRewardsNoCommission, totalDelegateStake)
+      totalRewardsPerYear <- calculateAverageRewardOverAYear(avgRewardAmount, emConfig.epochsPerYear)
+
+    } yield
+      RewardsInfo(
+        epochsPerYear = emConfig.epochsPerYear,
+        currentDagPrice = currentPrice,
+        nextDagPrice = nextPrice,
+        totalDelegatedAmount = totalDelegateStake,
+        latestAverageRewardPerDag = avgRewardAmount,
+        totalDagAmount = currentTotalSupply,
+        totalRewardPerEpoch = latestDelegateRewardsNoCommission,
+        totalRewardsPerYearEstimate = totalRewardsPerYear
+      )
+
+  private def processDelegations(info: GlobalSnapshotInfo): F[(Amount, Amount, Amount, Amount)] = {
+    val activeDelegatedStakes = info.activeDelegatedStakes
+      .getOrElse(SortedMap.empty[Address, List[DelegatedStakeRecord]])
+
+    val pendingWithdrawals = info.delegatedStakesWithdrawals
+      .getOrElse(SortedMap.empty[Address, List[PendingDelegatedStakeWithdrawal]])
+
+    val totalSpendableSupply = info.balances.values.map(_.value.value).sum
+    val totalPendingSupply = pendingWithdrawals.values.flatten.map(_.rewards.value.value).sum
+
+    val (totalStakeLocked, totalActiveRewards) = activeDelegatedStakes.values.flatten.foldLeft((0L, 0L)) {
+      case ((stakeAcc, rewardsAcc), record) =>
+        val stakeAmount = record.event.value.amount.value.value
+        val rewardsAmount = record.rewards.value
+        (stakeAcc + stakeAmount, rewardsAcc + rewardsAmount)
+    }
+
+    (
+      toAmount(totalStakeLocked),
+      toAmount(totalActiveRewards),
+      toAmount(totalStakeLocked + totalActiveRewards), // totalDelegateStake
+      toAmount(totalSpendableSupply + totalPendingSupply + totalActiveRewards) // currentTotalSupply
+    ).mapN(Tuple4.apply)
+  }
+
+  private def getLatestDelegateRewardTotal(snapshot: GlobalIncrementalSnapshot, info: GlobalSnapshotInfo): F[Amount] = {
+    val delegateRewards = snapshot.delegateRewards
+      .getOrElse(SortedMap.empty[PeerId, Map[Address, Amount]])
+
+    val nodeParams = info.updateNodeParameters
+      .getOrElse(SortedMap.empty[ID.Id, (Signed[UpdateNodeParameters], SnapshotOrdinal)])
+
+    val calcFullReward: (Long, (PeerId, Map[Address, Amount])) => Long = {
+      case (acc, (peerId, rewards)) =>
+        val nodeCommission = BigDecimal(
+          nodeParams.get(peerId.toId).map(_._1.delegatedStakeRewardParameters.reward).getOrElse(0.0)
+        )
```

Comment:
Similar issue here as above, where this method (or conversion,) should be associated with the class itself, such that we get back a decimal related to the class instance (and can therefore see how it was originally declared.

I'm also wondering if this should just be wrapped in the other fraction type which does the conversion there. It doesn't feel great having decimal logic encapsulated in places like this.